### PR TITLE
implemented registration of CRDs

### PR DESCRIPTION
**Pull request by dmitsh**
In the existing setup, each job submission requires us to define the resource's GVR (Group, Version, Resource), template, name format, and the format for pod names. These parameters remain constant.

Currently, if we have multiple "SubmitObj" tasks, we must repeat these parameters for each task.

I suggest we introduce a new task type called "RegisterObj." This task would store all the necessary parameters for a specific CRD in memory. When a job involving that CRD is submitted, the stored parameters would automatically be applied.

### CODE REVIEW

**Comment by lalitadithya** [comment]
@dmitsh , please correct me if I'm wrong, based on my understand and our older discussion, the reason why we need this change is because we cannot infer the `schema.GroupVersionResource` from the group, version, kind from the YAML. The reason being that the resource does not necessarily match the kind, i.e., the kind can be "Job", but the resource will be "jobs". Is this understanding correct? 

I recently leant about kubectl's api-resources, and I think this might be helpful here. If we run `kubectl api-resources --api-group=batch.volcano.sh -o wide`, we get the following output from which we can get the resource (jobs) from the kind (Job).
```
NAME   SHORTNAMES   APIVERSION                  NAMESPACED   KIND   VERBS
                     CATEGORIES
jobs   vcjob,vj     batch.volcano.sh/v1alpha1   true         Job    delete,deletecollection,get,list,patch,create,update,watch
```
We can parse the YAML to fetch the group and version (split the api version by `/` to get the group and version) and then we can subsequently use this information along with the [discovery package](https://github.com/kubernetes/client-go/tree/master/discovery) to fetch the required mapping. This would mean that the user will no longer have to specify the group version resource in the job submission schema. 

What do you think?

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -93,61 +95,62 @@ func (eng *Eng) RunTask(ctx context.Context, cfg *config.Task) error {
 
 // GetTask initializes and validates task
 func (eng *Eng) GetTask(cfg *config.Task) (Runnable, error) {
-	eng.mutex.Lock()
-	defer eng.mutex.Unlock()
+	//eng.mutex.Lock()
+	//defer eng.mutex.Unlock()
```
Comment (lalitadithya) [inline pkg/engine/engine.go:99]: Is this change intentional? Do we no longer required a mutex?

Comment (dmitsh) [reply pkg/engine/engine.go:99]: This is temporary, I will revert it back.

**Comment by dmitsh** [comment]
> @dmitsh , please correct me if I'm wrong, based on my understand and our older discussion, the reason why we need this change is because we cannot infer the `schema.GroupVersionResource` from the group, version, kind from the YAML. The reason being that the resource does not necessarily match the kind, i.e., the kind can be "Job", but the resource will be "jobs". Is this understanding correct?
> 
> I recently leant about kubectl's api-resources, and I think this might be helpful here. If we run `kubectl api-resources --api-group=batch.volcano.sh -o wide`, we get the following output from which we can get the resource (jobs) from the kind (Job).
> 
> ```
> NAME   SHORTNAMES   APIVERSION                  NAMESPACED   KIND   VERBS
>                      CATEGORIES
> jobs   vcjob,vj     batch.volcano.sh/v1alpha1   true         Job    delete,deletecollection,get,list,patch,create,update,watch
> ```
> 
> We can parse the YAML to fetch the group and version (split the api version by `/` to get the group and version) and then we can subsequently use this information along with the [discovery package](https://github.com/kubernetes/client-go/tree/master/discovery) to fetch the required mapping. This would mean that the user will no longer have to specify the group version resource in the job submission schema.
> 
> What do you think?

What if you have more than one CR from the same group?

File: pkg/engine/engine.go (hunk 1 above)
Comment (dmitsh) [reply pkg/engine/engine.go]: reverted.

File: pkg/engine/register_object_task.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	task.objTpl, err = template.ParseFiles(task.Template)
+	if err != nil {
+		return fmt.Errorf("%s: failed to parse template %s: %v", task.ID(), task.Template, err)
+	}
+
+	if len(task.NameFormat) == 0 {
+		return fmt.Errorf("%s: must specify name format for multiple object submissions", task.ID())
+	}
+
+	if len(task.PodnameFormat) != 0 {
```
Comment (lalitadithya) [inline pkg/engine/register_object_task.go:102]: nit:
```suggestion
	if len(task.PodNameFormat) != 0 {
```

Comment (dmitsh) [reply pkg/engine/register_object_task.go]: done

File: pkg/engine/check_pod_task.go [hunk 3]
```diff
@@ -209,24 +230,44 @@ func (task *CheckPodTask) verifyLabels(ctx context.Context, pod *v1.Pod) error {
...
+			}
 		}
-		if sz := podMap.Delete(pod.Name); sz == 0 {
-			task.log.Info("Accounted for all pods")
-			errs <- nil
-			return
+	}
+}
+
+func exp2regexp(expr []string) ([]*regexp.Regexp, error) {
```
Comment (lalitadithya) [inline pkg/engine/check_pod_task.go:264]: nit:

```suggestion
func (task *CheckPodTask) exp2regexp(expr []string) ([]*regexp.Regexp, error) {
```

or maybe put it a util pkg?

**Review by lalitadithya — APPROVED** [review]

File: pkg/engine/check_pod_task.go (hunk 3 above)
Comment (dmitsh) [reply pkg/engine/check_pod_task.go]: thanks! Moved to utils, added unit test.

**Review by lalitadithya — APPROVED** [review]
